File: Lecture7/RandPerm.java
import java.util.Random;
import java.util.Scanner;

public class RandPerm {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Random random = new Random();
        System.out.print("n?  ");
        int n = scanner.nextInt();

        int[] arr = new int[n];

        boolean[] isGenerated = new boolean[n]; // default : false

        for ( int i = 0; i < n; i++ ) {
            while ( true ){
                int temp = random.nextInt(n);
                if ( !isGenerated[temp] ) {
                    arr[i] = temp + 1;
                    isGenerated[temp] = true;
                    break;
                }
            }
        }

        for (int v : arr) {
            System.out.print(v + " ");
        }
        System.out.println();
    }
}

File: Lecture7/RandPerm.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(void){
    srand(time(NULL));
    printf("n?  ");
    int n; scanf("%d", &n);

    int* arr = (int*)malloc(sizeof(int)*n);

    int* isGenerated = (int*)malloc(sizeof(int)*n);
    for ( int i = 0; i < n; i++ ) {
        isGenerated[i] = 0;
    }

    for ( int i = 0; i < n; i++ ) {
        while ( 1 ){
            int temp = rand() % n;
            if ( !isGenerated[temp] ) {
                arr[i] = temp + 1;
                isGenerated[temp] = 1;
                break;
            }
        }
    }

    for (int i = 0; i < n; i++ ) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    free(arr);
    free(isGenerated);

    return 0;
}
